### ISSUE TITLE
** Uncaught Exception! ** when running "mvn package" on OSX

### ISSUE THREAD

**Issue by jchyip** [itdelatrisu/opsu#87]
**Version:** 0.9.0
**Build date:** 2015-06-03 19:54
**OS:** Mac OS X (x86_64)
**JRE:** 1.8.0_31
**Error:** `** Uncaught Exception! **`
**Stack trace:**
```
java.lang.UnsatisfiedLinkError: Can't load library: /Users/jyip/dev/opsu/./target/natives/liblwjgl.dylib
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1825)
	at java.lang.Runtime.load0(Runtime.java:809)
	at java.lang.System.load(System.java:1083)
	at org.lwjgl.Sys$1.run(Sys.java:70)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.lwjgl.Sys.doLoadLibrary(Sys.java:66)
	at org.lwjgl.Sys.loadLibrary(Sys.java:95)
	at org.lwjgl.Sys.<clinit>(Sys.java:112)
	at org.lwjgl.opengl.Display.<clinit>(Display.java:135)
	at org.newdawn.slick.AppGameContainer$1.run(AppGameContainer.java:39)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.newdawn.slick.AppGameContainer.<clinit>(AppGameContainer.java:36)
	at itdelatrisu.opsu.Opsu.main(Opsu.java:161)
```

**Comment by Bigpet** [comment]
Did you build that with maven? And do the jar releases work for you?

**Comment by jchyip** [comment]
"mvn install" triggers this, as does "mvn compile exec:exec"

"mvn install -Djar" works fine

On 3 June 2015 at 16:06, Peter Tissen <notifications@github.com> wrote:

> Did you build that with maven? And do the jar releases work for you?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/itdelatrisu/opsu/issues/87#issuecomment-108598954>.
>



-- 
Jason Che-han Yip
http://jchyip.blogspot.com
"The truth does not change according to our ability to stomach it."
(Flannery O'Connor)

**Comment by itdelatrisu** [comment]
Does the `liblwjgl.dylib` file exist?

**Comment by chong601** [comment]
well, ```mvn install``` will definitely give an error because it can't find the library file in ```Users/jyip/dev/opsu/./target/natives/liblwjgl.dylib``` since that folder is your opsu! project folder (those libs are kept in ```Users/.m2/repository```). Maven expects to find the libraries in the target folder by default **UNLESS** if you inform Maven the library iirc.
the use of "mvn install -Djar" is do the following: 
- Resource copy to project target folder
- Compile source code to .jar
- Copy all dependencies .jar file to project target folder
- use JarSplicePlus (itdelatrisu modified JarSplice)to shade (include) every single dependencies .jar file into another .jar (afaik opsu-<version>-shaded.jar).
- the final opsu-<version>-runnable.jar

**Comment by jchyip** [comment]
In that case, is there a documentation change required for README.md?  As in, is there another step required before mvn compile exec:exec would be expected to work?

**Comment by chong601** [comment]
you do need to have all required libraries added to opsu! dependencies.
```compile exec:exec``` does the same thing as install -Djar (except that it just compiles the source code and run directly in the ```target``` folder without packaging to .jar). 
if you can do ```install -Djar``` goal, you should be able to do ```compile exec:exec``` goal

**Comment by stdex** [comment]
Add to readme information about maven.
Need to use maven3 to compile project.
On ubuntu 14.04 this is something like:
>sudo apt-get purge maven maven2 maven3
>sudo apt-add-repository ppa:andrei-pozolotin/maven3
>sudo apt-get update
>sudo apt-get install maven3
>cd {project_dir}
>mvn compile exec:exec

**Comment by itdelatrisu** [comment]
@jchyip: does the post above solve your problem?

**Comment by jchyip** [comment]
No it does not.  I'm already using Maven 3.  Tried this on Windows and it works fine.  I don't see target/natives/liblwjgl.dylib in either OSX or Windows.  Are you saying that this is supposed to be from the Maven repo?

**Comment by itdelatrisu** [comment]
Oh, my bad, the file doesn't exist.  Are you using Java 8?  From [this post](https://github.com/MultiMC/MultiMC5/issues/371), it seems like this only affects Java 8 in OSX.  Not sure what to do...

**Comment by jchyip** [comment]
Looks like the short-term solution is to rename
target/native/liblwjgl.jnilib to target/native/liblwjgl.dylib.  Perhaps
just make a note of this for OSX / Java 8 users?

On 14 June 2015 at 11:05, Jeffrey Han <notifications@github.com> wrote:

> Oh, my bad, the file doesn't exist. Are you using Java 8? From this post
> <https://github.com/MultiMC/MultiMC5/issues/371>, it seems like this only
> affects Java 8 in OSX. Not sure what to do...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/itdelatrisu/opsu/issues/87#issuecomment-111838759>.
>



-- 
Jason Che-han Yip
http://jchyip.blogspot.com
"The truth does not change according to our ability to stomach it."
(Flannery O'Connor)

**Comment by itdelatrisu** [comment]
Well, it's possible to rename things in Maven ([link](http://stackoverflow.com/questions/3998596)) but I don't know if it's worth the trouble...

**Comment by itdelatrisu** [comment]
Seems like this was fixed when I updated LWJGL from 2.9.1 to 2.9.3.  This is included in the [changelog for the 2.9.2 release](http://legacy.lwjgl.org/changelogs/2.9.2-changelog.txt):
```
2014-11-28  Ioannis Tsakpinis <iotsakp@gmail.com>

	* src/java/org/lwjgl/LWJGLUtil.java, src/java/org/lwjgl/Sys.java,
	src/java/org/lwjgl/util/applet/AppletLoader.java: Replace .jnilib
	with .dylib on OS X with JDK 6. Resolve #100
```
